Add alerting rules for Promscale. (#1181)
timescale/promscale: docs/promscale_alerting.md

# Promscale Alerting

This file contains alerting rules which are symptom based. We recommend monitoring your Promscale deployments
and include the following rules as part of your alerting environment.

## Rules

> Copy the below in a `promscale_alerts.yaml` file.

Reviewer: This is all I need to do? :smile:

Author: yeah, i forgot to write adding this to prometheus.yml. Adding that.

# Promscale Alerting

This file contains alerting rules which are symptom based. We recommend monitoring your Promscale deployments
and include the following rules as part of your alerting environment.

## Rules

Copy the below in a `promscale_alerts.yaml` file.

```yaml
groups:

> - name: Promscale

Reviewer 2: ```suggestion
- name: promscale-general
```

  rules:
  - alert: PromscaleDown
    expr: absent(up{job=~".*promscale.*"})
    for: 2m
    labels:
      severity: critical
      component: connector
    annotations:
      summary: Promscale is down
      description: No Promscale instance was found.

> # Ingest

Reviewer 2: ```suggestion
- name: promscale-ingest
```

Instead of using comments in yaml code (which are lost in prometheus), we can put those alerts in separate groups.

Added bonus of doing this would be parallelization of alert expression execution in prometheus (groups are run in parallel, the content of a group is run sequentially).

  - alert: PromscaleIngestHighErrorRate
    expr: |
      (
        sum by (job, instance, type) (
          rate(promscale_ingest_requests_total{code=~"5.."}[5m])
        )
      /
        sum by (job, instance, type) (
          rate(promscale_ingest_requests_total[5m])
        )
      ) > 0.05

>     for: 30s # Assuming one scrape interval. 

Reviewer 2: This wasn't in the DD and shouldn't be put here. There is no reason to wait for this alert.

    labels:
      severity: warning
      component: connector
    annotations:
      summary: High error rate in Promscale ingestion

>       description: "Promscale ingestion is having a {{ $value }} error rate."

Reviewer 2: ```suggestion
      description: "Promscale ingestion is having a {{ $value | humanizePercentage }} error rate."
```

  - alert: PromscaleIngestHighErrorRate
    expr: |
      (
        sum by (job, instance, type) (
          rate(promscale_ingest_requests_total{code=~"5.."}[5m])
        )
      /
        sum by (job, instance, type) (
          rate(promscale_ingest_requests_total[5m])
        )
      ) > 0.1
    for: 30s
    labels:
      severity: critical
      component: connector
    annotations:
      summary: High error rate in Promscale ingestion

>       description: "Promscale ingestion is having a {{ $value }} error rate."

Reviewer 2: ```suggestion
      description: "Promscale ingestion is having a {{ $value | humanizePercentage }} error rate."
```

  - alert: PromscaleIngestHighLatency
    expr: |
      (
        histogram_quantile(
          0.90,
          sum by (job, instance, type, le) (
            rate(promscale_ingest_duration_seconds_bucket[5m])
          )
        ) > 10
      and
        sum by (job, instance, type) (
            rate(promscale_ingest_duration_seconds_bucket[5m])
        )
      ) > 0
    for: 1m
    labels:
      severity: warning
      component: connector
    annotations:
      summary: Slow Promscale ingestion
      description: "90% of ingestion batch took {{ $value }} seconds to ingest."
  - alert: PromscaleIngestHighLatency
    expr: |
      (
        histogram_quantile(
          0.90,
          sum by (job, instance, type, le) (
            rate(promscale_ingest_duration_seconds_bucket[5m])
          )
        ) > 30
      and
        sum by (job, instance, type) (
            rate(promscale_ingest_duration_seconds_bucket[5m])
        )
      ) > 0
    for: 1m
    labels:
      severity: critical
      component: connector
    annotations:
      summary: Slow Promscale ingestion
      description: "90% of ingestion batch took {{ $value }} seconds to ingest."

> # Query

Reviewer 2: ```suggestion
- name: promscale-query
```

same as above.

  - alert: PromscaleQueryHighErrorRate
    expr: |
      (
        sum by (job, instance, type) (
          rate(promscale_query_requests_total{code=~"5.."}[5m])
        )
      /
        sum by (job, instance, type) (
          rate(promscale_query_requests_total[5m])
        )
      ) > 0.05
    for: 1m
    labels:
      severity: warning
      component: connector
    annotations:
      summary: High error rate in querying Promscale

>       description: "Evaluating queries via Promscale has {{ $value }} error rate."

Reviewer 2: ```suggestion
      description: "Evaluating queries via Promscale has {{ $value | humanizePercentage }} error rate."
```

  - alert: PromscaleQueryHighErrorRate
    expr: |
      (
        sum by (job, instance, type) (
          rate(promscale_query_requests_total{code=~"5.."}[5m])
        )
      /
        sum by (job, instance, type) (
          rate(promscale_query_requests_total[5m])
        )
      ) > 0.1
    for: 1m
    labels:
      severity: critical
      component: connector
    annotations:
      summary: High error rate in querying Promscale

>       description: "Evaluating queries via Promscale had {{ $value }} error rate."

Reviewer 2: ```suggestion
      description: "Evaluating queries via Promscale had {{ $value | humanizePercentage }} error rate."
```

  - alert: PromscaleQueryHighLatency
    expr: |
      (
        histogram_quantile(
          0.90,
          sum by (job, instance, type, le) (
            rate(promscale_query_duration_seconds_bucket[5m])
          )
        ) > 5
      and
        sum by (job, instance, type) (
          rate(promscale_query_duration_seconds_bucket[5m])
        ) > 0
      )
    for: 1m
    labels:
      severity: warning
      component: connector
    annotations:
      summary: Slow Promscale querying
      description: "90% of the queries took {{ $value }} seconds to evaluate."
  - alert: PromscaleQueryHighLatency
    expr: |
      (
        histogram_quantile(
          0.90,
          sum by (job, instance, type, le) (
            rate(promscale_query_duration_seconds_bucket[5m])
          )
        ) > 10
      and
        sum by (job, instance, type) (
          rate(promscale_query_duration_seconds_bucket[5m])
        ) > 0
      )
    for: 1m
    labels:
      severity: critical
      component: connector
    annotations:
      summary: Slow Promscale querying
      description: "90% of the queries took {{ $value }} seconds to evaluate."

> # Cache

Reviewer 2: ```suggestion
- name: promscale-cache
```

same as above.

  - alert: PromscaleCacheHighNumberOfEvictions
    expr: |
      (
        sum by (job, instance, name, type) (
          rate(promscale_cache_evictions_total[5m])
        )
      /
        sum by (job, instance, name, type) (
          promscale_cache_capacity_elements
        )
      ) > 0.2
    for: 5m
    labels:
      severity: warning
      component: connector
    annotations:
      summary: High cache eviction in Promscale

>       description: "Promscale clockcache is evicting at {{ $value }} entries a second."

Reviewer 2: ```suggestion
      description: "Promscale {{ $commonLabels.name }} is evicting at {{ $value }} entries a second."
```

Let's be future-proof and use templates for the cache name which is problematic.

  - alert: PromscaleCacheTooSmall
    expr: |
      (
        sum by (job, instance, type, name) (
          rate(promscale_cache_query_hits_total[5m])
        )
      /
        sum by (job, instance, type, name) (
          rate(promscale_cache_queries_total[5m])
        )
      ) < 0.9
    for: 5m
    labels:
      severity: warning
      component: connector
    annotations:
      summary: High cache eviction in Promscale
      description: "Promscale clockcache has a hit ratio of {{ $value }}."

> ```

Reviewer 2: Putting this content in an `md` file allows users only to copy it manually without any ability to use this in an automated fashion. Instead, let's follow how other projects include such content by doing the following:
1. Create a directory structure as follows:
```
docs
└── mixin
    ├── alerts
    │   └── alerts.yaml
    ├── dashboards
    │   └── dashboard.yaml
    └── README.md
```
2. Put the content from the highlighted section in `alerts.yaml`
3. Put the instructions on how to use this in `README.md` so it is easily readable when going to the directory via GitHub UI

Such structure will allow us to easily extend content and in the future potentially include it in https://monitoring.mixins.dev/

Side note: This such directory schema will create a [monitoring mixin](https://github.com/monitoring-mixins/docs/blob/master/design.pdf).
